File: Problems/143_reorderList.ts
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

function reorderList(head: ListNode | null): void {
    if (head == null || head.next == null || head.next.next == null) return;

    //get 'second half' of list in slow
    let fast = head;
    let slow = head;
    while (fast.next != null && fast.next.next != null) { //fast is the foresight, to ensure slow only goes upto the middle/previous middle
        fast = fast.next.next;
        slow = slow.next;
    }

    //reverse slow
    let storage = slow.next;
    slow.next = null;
    slow = null;
    while (storage != null) {
        const temp = storage.next;
        storage.next = slow;
        slow = storage;
        storage = temp;
    }

    //'interlace' slow with head
    while (slow != null) {
        const temp = head.next;
        head.next = slow;
        head = head.next;
        slow = temp;
    }

    //we lose track of the pointer head, however the original variable has now been modified
};

const testList1 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, null))));
const testList2 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, null)))));
console.log(reorderList(testList2));
console.log(testList2);

File: Problems/143_reorderList.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: None Do not return anything, modify head in-place instead.
        """
        
        if head == None or head.next == None: #length <= 1, so we can just return head
            return

        #find middle of list using fast and slow pointer; slow pointer moves 1 space for each time fast pointer moves 2 spaces, thus meaning slow pointer is at middle when 
        fast = head
        slow = head
        while fast != None and fast.next != None:
            fast = fast.next.next
            slow = slow.next

        #slow is second half of list; reverse slow
        storage = slow.next
        slow.next = None #this causes us to lose data on head; which is not wanted; however we are only losing items after the first half which is all we want head for anyway, so it won't matter for this problem
        #losing all items after and including slow.next is actually beneficial since it means we don't have to check separately for odd and even cases, since the final item added will always have None for next

        while storage != None:
            temp = storage.next
            storage.next = slow
            slow = storage
            storage = temp

        current = head
        temp = slow
        while current.next != None: #once we have swapped to the final item in slow, the current.next will be None and so the loop will break; use slow as temp
            current.next, temp = temp, current.next
            current = current.next

        #head is controlled by current, which should now be correct
    
solution = Solution()
lst = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5, None)))))
solution.reorderList(lst)

print(lst.val)
print(lst.next.val)
print(lst.next.next.val)
print(lst.next.next.next.val)
print(lst.next.next.next.next.val)